modules/seqexec/engine/src/main/scala/seqexec/engine/Action.scala

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2016-2019 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package seqexec.engine
+
+import fs2.Stream
+import seqexec.engine.Result.{Error, PartialVal, PauseContext, RetVal}
+import seqexec.model.ActionType
+import seqexec.model.enum.ActionStatus
+import monocle.macros.Lenses
+
+@Lenses
+final case class Action[F[_]](
+  kind: ActionType,
+  gen: Stream[F, Result[F]],
+  state: Action.State[F]
+)
+
+object Action {
+
+  final case class State[F[_]](runState: ActionState[F], partials: List[PartialVal])
+
+  sealed trait ActionState[+F[_]] {
```

Comment:
On the ADT for `ActionState` only some of them have a need for an `F`. With the `[+F[_]]` I can do
`case object Started extends ActionState[Nothing]`
Or maybe I'm confused

Reply:
Oh I see, it's a GADT. Yeah that's fine.

Reply:
Here is a GADT. Note that some constructors instantiate the type variable `A` at specific types. That's what a GADT lets you do. Otherwise every case would have to be generic in `A`.

```scala
sealed trait Stuff[A]
case class  Foo[A](a: A)   extends Stuff[A]
case class  Bar(d: Double) extends Stuff[String] // <--
case object Baz            extends Stuff[Int]    // <--
```

So what does this let you do? It lets you write a method that's generic in `A` but by pattern-matching you can refine your knowledge about what `A` is.

```scala
def whatever[A](sa: Stuff[A]): A =
  sa match {
    case Foo(a) => a           // this branch returns A
    case Bar(d) => d.toString  // this branch returns String (!)
    case Baz    => 42          // this branch returns Int    (!)
  }
```

In fact in the `Bar` case you could write `implicitly[A =:= String]` and it would compile because it's true, but only to the right of the `=>`. So you can think of pattern-matching on a GADT as narrowing both the data constructor (we know it's a `Bar`) *and* narrowing the type argument (we also know that `A =:= String`).

Reply:
So this would be true?
`whatever(Foo(2)) + 2 == 4`

Reply:
```scala
@ whatever(Foo(2)) + 2 == 4 
res2: Boolean = true
```